GlobalConstantsService/src/ParticleDataList.cc

```diff
@@ -0,0 +1,121 @@
+
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+
+#include "cetlib_except/exception.h"
+
+#include "Offline/GlobalConstantsService/inc/ParticleDataList.hh"
+#include "Offline/ConfigTools/inc/ConfigFileLookupPolicy.hh"
+#include "Offline/ConfigTools/inc/SimpleConfig.hh"
+#include "Offline/DataProducts/inc/PDGCode.hh"
+
+
+namespace mu2e {
+
+  // *************************************************************
+
+  ParticleDataList::ParticleDataList( SimpleConfig const& config ) {
+
+    ConfigFileLookupPolicy findConfig;
+    
+    std::string tableFilename = findConfig(
+                    config.getString("particleDataList.filename",
+                   "Offline/GlobalConstantsService/data/ParticleList.txt") );
+    if( tableFilename.empty() ) {
+      throw cet::exception("PARTICLELIST_NO_FILE") 
+        << "ParticleDataList: file name not found\n";
+    }
+
+    std::ifstream in(tableFilename.c_str());
+    if ( !in ) {
+      throw cet::exception("PARTICLELIST_OPEN_FAIL")
+          << "Unable to open particle data file: "
+          << tableFilename << "\n";
+    }
+
+    std::string line;
+    std::string str;
+    std::vector<std::string> words;
+    while ( std::getline(in,line) ) {
+      std::istringstream iss(line);
+      while (iss >> str) {
+        words.emplace_back(str);
+      }
+      if ( words.size() != 7 ) {
+        throw cet::exception("PARTICLELIST_BAD_LINE")
+          << "ParticleList: Not 7 words on line: "
+          << line << "\n";
+      }
+      int id = std::stoi(words[0]);
+      _list.try_emplace(id, id, words[1], words[2], 
+         std::stof(words[4]), std::stof(words[5]), std::stof(words[6]) );
+      _names.try_emplace(words[1],id);
+      if( words[2] != "none" ) _names.emplace(words[2],id);
+      if( words[3] != "none" ) _names.emplace(words[3],id);
+      words.clear();
+    }
+
+
+  }
+
+  // *************************************************************
+
+  const ParticleData& ParticleDataList::particle( int id ) const {
+
+    auto it = _list.find(id);
+    if( it != _list.end() ) {
+      return it->second;
+    }
+
+    if ( id <= PDGCode::G4Threshold ){    
+      throw cet::exception("PARTICLELIST_BAD_ID")
+        << "ParticleList: Failed to look up id: "
+        << id << "\n";
+    }
+
+    // this is a unknown geant nucleus, add it to the list
+
+    int pA = (std::abs(id)/10)%1000;
+    int pZ = (std::abs(id)/10000)%1000;
+
+    // the nuclear excitation level
+    int exc = id%10;
+    float protonMass = particle(PDGCode::p_plus).mass();
+    float mass = (pA+exc)*protonMass;
+
+    std::ostringstream pName;
+    pName << "Mu2e_" << id;
+    std::string name = pName.str();
+
+    _list.try_emplace(id,id, name, name, float(pZ), mass, 0.0);
+
+    return _list.at(id);
+    
```

Comment:
Do we want a smarter error message in the case that try_emplace fails?  As written you will just get the out of bounds message from _list.

Reply:
if it finds the key, it will not change anything. If it does not find the key, it will construct the pair.  Either way, the key will be there.   The only failure might be something like running out of memory.  What error are you thinking we need to catch?

Reply:
You are right - there is no real use case.  If it runs out of memory it will core dump and we won't be able to issue a well phrased warning.